mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -745,4 +763,27 @@ private[spark] class CosineDistanceMeasure extends DistanceMeasure {
   override def distance(v1: VectorWithNorm, v2: VectorWithNorm): Double = {
     1 - dot(v1.vector, v2.vector) / v1.norm / v2.norm
   }
+
+  /**
+   * Updates the value of `sum` adding the `point` vector.
+   * @param point a `VectorWithNorm` to be added to `sum` of a cluster
+   * @param sum the `sum` for a cluster to be updated
+   */
+  override def updateClusterSum(point: VectorWithNorm, sum: Vector): Unit = {
+    axpy(1.0 / point.norm, point.vector, sum)
```

Comment:
do we need to ignore zero points here?

Reply:
the cosine similarity/distance is not defined for zero points: if there were 0 points we would have earlier failures while computing any cosine distance involving them.

Reply:
In scala, `1.0 / 0.0` generate `Infinity`, what about directly throw an exception instead?

Reply:
Thanks, I agree. I added an assertion before computing the cosine distance and a test case for this situation. Thank you for your comment.